### PR TITLE
[Converged MME] Standalone spgw

### PR DESCRIPTION
**Pull request by ssanadhya**
## Summary
This change set adds support for a simple SPGW in AGW, which uses gtpv1u datapath through libgtpnl and performs local IP allocation. Specifically, this change set
- Adds a flag to enable/disable the communication between MME and Sessiond or Mobilityd in AGW
- When Sessiond and Mobilityd are disabled, the SPGW task within MME uses the libgtpnl datapath instead of OVS, does not send any requests to PCEF and performs IP allocation through a pre-configured block of IPs.

This creates a simple EPC process in MME, which can be run without any OVS dependencies.

## Test Plan

- Feature `agw_of` tested with S1ap integration tests
- Feature `agw_gtpnl` tested in local setup by running setup_gw2_gtp.sh on magma dev VM and setup_gw2_s1tests.sh on magma test VM, then make integ_test

## Commits included from converged-mme repo
1796fb889       Shruti Sanadhya Thu Jul 16 00:01:21 2020 -0700  Link Gtpv1u task to openflow library only when needed
7bd3250b9       Shruti Sanadhya Thu Jul 9 09:34:03 2020 -0700   Fixing log message in gw2 setup script
d6eea0f3a       Shruti Sanadhya Fri Jun 12 13:00:54 2020 -0700  Adding SGI interface switch to setup script
498ea943b       Shruti Sanadhya Wed May 13 21:57:06 2020 -0700  Adding checks in libgtpnl install and removing unused Magma services
083848d56       Shruti Sanadhya Wed May 13 15:12:17 2020 -0700  Adding success check in setup script for GW2
bdbb5a125       rashmi  Fri Apr 17 21:04:41 2020 +0530  Incorporated review comments
c7c184adf       rashmi  Fri Apr 17 20:56:10 2020 +0530  Adding code to allocate UE IP adress locally UE pool is configured in spgw.conf.template
1b87c854b       rashmi  Fri Apr 17 20:23:06 2020 +0530  Picked the changes from branch, gtpnl_datapath with commit ID, cc080f904bcd51626aff523652f70ccd83e3e627

### CODE REVIEW

File: lte/gateway/c/oai/tasks/sgw/sgw_task.c [hunk 1]
```diff
@@ -117,6 +117,7 @@ static int handle_message (zloop_t* loop, zsock_t* reader, void* arg)
 
     case SGI_UPDATE_ENDPOINT_RESPONSE: {
       sgw_handle_sgi_endpoint_updated(
+          spgw_state,
```
Comment (ulaskozat) [inline lte/gateway/c/oai/tasks/sgw/sgw_task.c:120]: is this formatted?

File: lte/gateway/c/oai/tasks/sgw/sgw_task.c [hunk 2]
```diff
@@ -130,6 +131,7 @@ static int handle_message (zloop_t* loop, zsock_t* reader, void* arg)
     case S11_NW_INITIATED_DEACTIVATE_BEARER_RESP: {
       //Handle Dedicated bearer deactivation Rsp from MME
       sgw_handle_nw_initiated_deactv_bearer_rsp(
+          spgw_state,
```
Comment (ulaskozat) [inline lte/gateway/c/oai/tasks/sgw/sgw_task.c:134]: looks couple of spaces off wrt the other arguments.

File: lte/gateway/c/oai/tasks/sgw/sgw_handlers.h [hunk 3]
```diff
@@ -66,8 +67,9 @@ int sgw_handle_nw_initiated_actv_bearer_rsp(
   const itti_s11_nw_init_actv_bearer_rsp_t* const s11_actv_bearer_rsp,
   imsi64_t imsi64);
```
Comment (ulaskozat) [inline lte/gateway/c/oai/tasks/sgw/sgw_handlers.h:68]: why do these calls not need spgw_state?

File: lte/gateway/c/oai/tasks/sgw/pgw_config.c [hunk 4]
```diff
@@ -105,6 +107,34 @@ int pgw_config_process(pgw_config_t* config_pP) {
...
+    }
+    uint32_t network_hbo = range_high_hbo & (UINT32_MAX << (32 - nbits));
+    uint32_t netmask_hbo = 0xFFFFFFFF << (32 - nbits);
+    config_pP->ue_pool_network[i].s_addr = htonl(network_hbo);
+    config_pP->ue_pool_netmask[i].s_addr = htonl(netmask_hbo);
+
+    // Any math should be applied onto host byte order i.e. ntohl()
+    each_ip_addr.s_addr = config_pP->ue_pool_range_low[i].s_addr;
+    while (htonl(each_ip_addr.s_addr) <=
+           htonl(config_pP->ue_pool_range_high[i].s_addr)) {
```
Comment (ulaskozat) [inline lte/gateway/c/oai/tasks/sgw/pgw_config.c:129]: isn't `config_pP->ue_pool_range_high[i].s_addr` already in network byte order?

Comment (ulaskozat) [reply lte/gateway/c/oai/tasks/sgw/pgw_config.c:129]: same with `each_ip_addr.s_addr`, it is copied from `config_pP->ue_pool_range_low[i].s_addr` and looks like it is also network byte order.

Comment (ulaskozat) [reply lte/gateway/c/oai/tasks/sgw/pgw_config.c:129]: Please double check whether these should be instead converted into host byte order as we are doing integer ranking and comparisons.

Comment (ulaskozat) [reply lte/gateway/c/oai/tasks/sgw/pgw_config.c:129]: In fact why don't we simply iterate with 
```
for (int i = range_low_hbo; i <= range_high_hbo; ++i) { 
  ip4_ref              = calloc(1, sizeof(conf_ipv4_list_elm_t)); 
  ip4_ref->addr.s_addr = htonl(i); 
  STAILQ_INSERT_TAIL(&config_pP->ipv4_pool_list, ip4_ref, ipv4_entries);
}
```

File: lte/gateway/c/oai/include/pgw_config.h [hunk 5]
```diff
@@ -114,9 +113,15 @@ typedef struct pgw_config_s {
   bool masquerade_SGI;
 
   int num_ue_pool;
-  uint8_t ue_pool_mask[PGW_NUM_UE_POOL_MAX];
-  struct in_addr ue_pool_addr[PGW_NUM_UE_POOL_MAX];
-
+#define PGW_NUM_UE_POOL_MAX 2
```
Comment (ulaskozat) [inline lte/gateway/c/oai/include/pgw_config.h:116]: why 2?

File: lte/gateway/c/oai/tasks/sgw/pgw_config.c [hunk 6]
```diff
@@ -306,37 +331,59 @@ int pgw_config_parse_file(pgw_config_t* config_pP) {
...
               } else {
-                OAILOG_ERROR(
-                    LOG_SPGW_APP, "CONFIG POOL ADDR IPV4: BAD MASQ: %d\n",
-                    prefix_mask);
+                OAI_FPRINTF_ERR(
+                    "CONFIG POOL ADDR IPV4: BAD ADRESS: %s\n",
+                    bdata(address_high));
+                return RETURNerror;
+              }
+              if (htonl(config_pP->ue_pool_range_low[config_pP->num_ue_pool]
```
Comment (ulaskozat) [inline lte/gateway/c/oai/tasks/sgw/pgw_config.c:370]: htonl -> ntohl ?

File: lte/gateway/c/oai/tasks/sgw/pgw_config.c [hunk 7]
```diff
@@ -306,37 +331,59 @@ int pgw_config_parse_file(pgw_config_t* config_pP) {
...
+                            .s_addr) >=
+                  htonl(config_pP->ue_pool_range_high[config_pP->num_ue_pool]
```
Comment (ulaskozat) [inline lte/gateway/c/oai/tasks/sgw/pgw_config.c:372]: htonl -> ntohl ?

File: lte/gateway/c/oai/tasks/sgw/pgw_config.c [hunk 8]
```diff
@@ -306,37 +331,59 @@ int pgw_config_parse_file(pgw_config_t* config_pP) {
...
+                            .s_addr)) {
+                OAI_FPRINTF_ERR(
+                    "CONFIG POOL ADDR IPV4: BAD RANGE: %s (%d %d)\n",
+                    bdata(range),
+                    htonl(config_pP->ue_pool_range_low[config_pP->num_ue_pool]
```
Comment (ulaskozat) [inline lte/gateway/c/oai/tasks/sgw/pgw_config.c:377]: htonl -> ntohl ?

File: lte/gateway/c/oai/tasks/sgw/pgw_config.c [hunk 9]
```diff
@@ -306,37 +331,59 @@ int pgw_config_parse_file(pgw_config_t* config_pP) {
...
+                              .s_addr),
+                    htonl(config_pP->ue_pool_range_high[config_pP->num_ue_pool]
```
Comment (ulaskozat) [inline lte/gateway/c/oai/tasks/sgw/pgw_config.c:379]: htonl -> ntohl ?

File: lte/gateway/c/oai/tasks/sgw/pgw_config.c [hunk 10]
```diff
@@ -306,37 +331,59 @@ int pgw_config_parse_file(pgw_config_t* config_pP) {
...
-            AssertFatal(2 == list->qty, "Bad CIDR address %s", bdata(cidr));
+              bsplit(range, PGW_CONFIG_STRING_IPV4_ADDRESS_RANGE_DELIMITER);
+            if (list->qty == 2) {
+              bstring address_low  = list->entry[0];
+              bstring address_high = list->entry[1];
 
-            address = list->entry[0];
-            mask    = list->entry[1];
+              btrimws(address_low);
+              btrimws(address_high);
```
Comment (ulaskozat) [inline lte/gateway/c/oai/tasks/sgw/pgw_config.c:348]: are these formatted?

**Comment by 119Vik** [comment]
Can one of the admins verify this patch?

**Comment by rdefosse** [comment]
MAGMA-OAI-TESTS

**Comment by ekowtaylor** [comment]
@ssanadhya, @ulaskozat  I'm trying to help close some of these older PRs. Since this is still in draft mode, are your planning to move forward with it or it can be closed?

**Comment by ulaskozat** [comment]
I think we can close this.

**Comment by ekowtaylor** [comment]
Thanks @ulaskozat, closing PR.

**Pull request closed without merging**